SharpToken/Lib/Internals/BytePairEncodingCore/BytePairEncodingCore.net8.cs

```diff
@@ -0,0 +1,129 @@
+#if NET8_0_OR_GREATER
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.RegularExpressions;
+
+
+namespace SharpToken;
+
+internal sealed class BytePairEncodingCore
+{
+    public BytePairEncodingCore(
+        BytePairIndex bytePairEncoder,
+        Dictionary<string, int> specialTokenEncoder,
+        Regex tokenPatternRegex
+    )
+    {
+        Encoder = bytePairEncoder;
+        Decoder = bytePairEncoder.ToDictionary(pair => pair.Value, pair => pair.Key);
+
+        SpecialTokensEncoder = specialTokenEncoder;
+        SpecialTokensDecoder = specialTokenEncoder.ToDictionary(pair => pair.Value, pair => Encoding.UTF8.GetBytes(pair.Key));
+        RegexTls = tokenPatternRegex;
+    }
+
+    public BytePairIndex Encoder { get; }
+    public Dictionary<string, int> SpecialTokensEncoder { get; }
+    public Dictionary<int, byte[]> Decoder { get; }
+    public Dictionary<int, byte[]> SpecialTokensDecoder { get; }
+    public Regex RegexTls { get; }
+
+    public (List<int> tokens, int count) EncodeNative(ReadOnlySpan<char> text, IReadOnlyCollection<string> allowedSpecialTokens, bool countOnly)
+    {
+        var encodedTokens = countOnly ? null : new List<int>((int) Math.Ceiling(text.Length / 4d) is var capacity && capacity > 4 ? capacity : 4);
+        var startIndex = 0;
+        var count = 0;
+        var pool = ArrayPool<byte>.Shared;
+
+        while (true)
+        {
+            var slice = text[startIndex..];
+
+            var nextSpecialMatch = allowedSpecialTokens.FindMatch(slice);
+            if (nextSpecialMatch.Success)
+            {
+                slice = slice[..nextSpecialMatch.Index];
+            }
+
+            foreach (var match in RegexTls.EnumerateMatches(slice))
+            {
```

Comment:
This `net8.0` specific implementation can not be shared with `net6.0` and `netstandard` because `Regex.EnumerateMatches(ReadOnlySpan<char>)` does not exist there. 

The other versions relay in `string` and can not use `ReadOnlySpan<char>` in Regex.